middleware/rate_limiter.go

```diff
@@ -0,0 +1,97 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+	"net/http"
+	"sync"
+)
+
+// TokenStore is the interface to be implemented by custom stores.
+type TokenStore interface {
+	// Stores for the rate limiter have to implement the ShouldAllow method
+	ShouldAllow(identifier string) bool
+}
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// SourceFunc uses echo.Context to extract the identifier for a visitor
+		SourceFunc func(context echo.Context) string
+		// Store defines a store for the rate limiter
+		Store TokenStore
+	}
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+}
+
+// RateLimiter returns a rate limiting middleware
+func RateLimiter(source func(context echo.Context) string, store TokenStore) echo.MiddlewareFunc {
+	config := DefaultRateLimiterConfig
+	config.SourceFunc = source
+	config.Store = store
+
+	return RateLimiterWithConfig(config)
+}
+
+// RateLimiterWithConfig returns a rate limiting middleware
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.SourceFunc == nil {
+		panic("Source function must be provided")
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier := config.SourceFunc(c)
+
+			allowed := config.Store.ShouldAllow(identifier)
+			if !allowed {
+				return c.JSON(http.StatusTooManyRequests, nil)
+			}
+			return next(c)
+		}
+	}
+}
+
+// InMemoryStore is the built-in store implementation for RateLimiter
+type InMemoryStore struct {
+	visitors map[string]*rate.Limiter
+	mutex    sync.Mutex
+	rate     rate.Limit
+	burst    int
+}
+
+// ShouldAllow implements TokenStore.ShouldAllow
+func (store *InMemoryStore) ShouldAllow(identifier string) bool {
+	store.mutex.Lock()
+	defer store.mutex.Unlock()
```

Comment:
Defer had performance issues at least in Go 1.12-1.13. Explicit unlock may be faster.
As this component is performance critical it might be worth the effort.

Reply:
Hmmm... Didn't know about that. Will convert this to an explicit Unlock at the end of the function then

Reply:
Yes, it was Go 1.14 that resolved it. See https://github.com/golang/go/issues/14939

Reply:
So not sure we need to change that, Go 1.12 is EOL and Go 1.13 will be next year. So I think we leave that as is.
You can benchmark if you like, would be interesting ;-)

Reply:
So I benchmarked the with the `defer` and it is slower coming in at 240 ns/op while without the defer it is at 227 ns/op

Not much of a loss but I think I'll just stick with this current impl without defer. 🥂

Reply:
Good to know! An educated decission now ;-)